[IRGen][DebugInfo] split method declaration and definition.

dianqk opened the pull request:
When the function is a method, we only want a `DW_AT_declaration`
there, because LLVM LTO can't unify type definitions when a child DIE is a
full subprogram definition. Now the subprogram definition gets added at the
CU level with a specification link back to the abstract declaration.

Both GCC and Clang write debuginfo this way for C++ class methods. [Now Rust is too](https://github.com/rust-lang/rust/pull/111167).

Related discussions:
- [[Verifier] definition subprograms cannot be nested within DICompositeType when enabling ODR.](https://reviews.llvm.org/D152095)
- https://github.com/llvm/llvm-project/issues/61932
- https://github.com/rust-lang/rust/pull/111167

Fixes #53940.

DianQK commented:
@swift-ci please test

adrian-prantl commented on test/DebugInfo/method.swift:
```diff
@@ -0,0 +1,27 @@
+// CHECK: distinct !DISubprogram(name: "static_method",
+// CHECK-SAME: declaration:
+// CHECK: distinct !DISubprogram(name: "method",
+// CHECK-SAME: declaration:
```
Could you also add a CHECK testing that the declaration is indeed for the same method and that the parent type (e.g., Foo) lists the declaration, not the definition?

DianQK commented:
@adrian-prantl  Could you help me invoke `swift-ci` and merge changes? This is my first pull request.

adrian-prantl commented:
@swift-ci test and merge

adrian-prantl commented:
@swift-ci test and merge

DianQK commented:
After adding the parameter `--param backtracing`, I can reproduce the `test/Backtracing/Crash.swift` error. Trying to solve it.

DianQK commented:
After this PR, a DWARF in the following format may be generated:
```
0x00000d58:       DW_TAG_subprogram
                    DW_AT_linkage_name	("$s8CrashOpt0A0V5$mainyyFZ")
                    DW_AT_name	("$main")
                    DW_AT_type	(0x00000f65 "$sytD")
                    DW_AT_declaration	(true)
                    DW_AT_external	(true)
                    DW_AT_APPLE_optimized	(true)
...
0x00000d9d:         DW_TAG_inlined_subroutine
                      DW_AT_abstract_origin	(0x000011d2 "$s8CrashOpt0A0V5$mainyyFZ")
                      DW_AT_ranges	(0x00000210
                         [0x00000000000011e0, 0x00000000000012f3)
                         [0x00000000000012fb, 0x0000000000001307))
                      DW_AT_call_file	("/home/dianqk/Projects/swift-project/swift/test/Backtracing/Crash.swift")
                      DW_AT_call_line	(0)
...
0x000011d2:   DW_TAG_subprogram
                DW_AT_specification	(0x00000d58 "$s8CrashOpt0A0V5$mainyyFZ")
                DW_AT_inline	(DW_INL_inlined)
```
I put [b00d1520f89bb7df1d2a3eeee57cae7682755634](https://github.com/apple/swift/pull/67077/commits/b00d1520f89bb7df1d2a3eeee57cae7682755634) to this PR to show on CI that this did work.

Opening a new PR should be a better option. #67592

DianQK commented:
I also tried to solve the failure of `lang/swift/async/stepping/step-in/task-switch/TestSwiftTaskSwitch.py` on macOS.
The https://github.com/apple/llvm-project/pull/7114 is a change that can pass the test case. But I have no lldb related knowledge.
This might be something to ask @kastiglione to take a look at, thanks.

al45tair commented:
I've reviewed the backtracing change in #67592. TL/DR: It looks fine, though I think I'd like it to check the tag to make sure the `DW_AT_specification` is pointing at the kind of thing we expect.

al45tair reviewed (approved):
Seems reasonable to me, modulo my comment about checking the tag.

al45tair commented:
@swift-ci Please test

DianQK commented:
I figured out why the `TestSwiftTaskSwitch.py` test case was failing. I tried to submit the patch https://reviews.llvm.org/D157184.

dianqk committed:
[IRGen][DebugInfo] Split method declaration and definition.

When the function is a method, we want a DW_AT_declaration there.
Because there's no good way to cross the CU boundary to insert a nested
DISubprogram definition in one CU into a type defined in another CU when
doing LTO builds.

DianQK commented:
Rebase and update code comments.

felipepiovezan commented:
@swift-ci test
